topic/types.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package topic
+
+import (
+	"github.com/m3db/m3cluster/client"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/services"
+)
+
+// Topic defines the topic of messages.
+type Topic interface {
+	// Name returns the name of the topic.
+	Name() string
+
+	// SetName sets the name of the topic.
+	SetName(value string) Topic
+
+	// NumberOfShards returns the total number of shards of the topic.
+	NumberOfShards() uint32
+
+	// SetNumberOfShards sets the total number of shards of the topic.
+	SetNumberOfShards(value uint32) Topic
+
+	// ConsumerServices returns the consumers of the topic.
+	ConsumerServices() []ConsumerService
+
+	// SetConsumerServices sets the consumers of the topic.
+	SetConsumerServices(value []ConsumerService) Topic
```

Comment:
Do we want to include an `AddConsumerService` method so that users dont have to pass in an entire list of all the existing consumer services if they just want to append one service to the end of a list. 

Also presumably this topic metadata is stored in KV, however from what I can tell this function does not take a version so we can not leverage check and set.

Reply:
The CAS is not an issue, when you do a get you will get the version and the topic, you manipulate the topic then you can check and set with old version you got. For AddConsumerService, I'm on the fence, I guess we can do it for both the add and remove cases, maybe I'll add it when I really need it. Right now I need the Set function only for tests.

Reply:
Added version anyways so the user of topic can log version if needed